contracts/service-registry/src/state.rs

```diff
@@ -130,16 +131,72 @@ pub enum AuthorizationState {
     Authorized,
 }
 
-// maps service_name -> Service
-pub const SERVICES: Map<&str, Service> = Map::new("services");
-// maps (service_name, chain_name, worker_address) -> ()
-pub const WORKERS_PER_CHAIN: Map<(&str, &ChainName, &Addr), ()> = Map::new("workers_per_chain");
-// maps (service_name, worker_address) -> Worker
-pub const WORKERS: Map<(&str, &Addr), Worker> = Map::new("workers");
+type ChainNames = HashSet<ChainName>;
+type ServiceName = str;
+type WorkerAddress = Addr;
+
+pub const SERVICES: Map<&ServiceName, Service> = Map::new("services");
+pub const WORKERS_PER_CHAIN: Map<(&ServiceName, &ChainName, &WorkerAddress), ()> =
+    Map::new("workers_per_chain");
+pub const CHAINS_PER_WORKER: Map<(&ServiceName, &WorkerAddress), ChainNames> =
+    Map::new("chains_per_worker");
+pub const WORKERS: Map<(&ServiceName, &WorkerAddress), Worker> = Map::new("workers");
+
+pub fn register_chain_support(
+    storage: &mut dyn Storage,
+    service_name: String,
+    chains: Vec<ChainName>,
+    worker: WorkerAddress,
+) -> Result<(), ContractError> {
+    CHAINS_PER_WORKER.update(storage, (&service_name, &worker), |current_chains| {
+        let mut current_chains = current_chains.unwrap_or_default();
+        current_chains.extend(chains.iter().cloned());
+        Ok::<HashSet<ChainName>, ContractError>(current_chains)
+    })?;
+
+    for chain in chains.iter() {
+        WORKERS_PER_CHAIN.save(storage, (&service_name, chain, &worker), &())?;
+    }
+
+    Ok(())
+}
+
+pub fn may_load_chains_per_worker(
+    storage: &dyn Storage,
+    service_name: String,
+    worker_address: WorkerAddress,
+) -> Result<HashSet<ChainName>, ContractError> {
+    CHAINS_PER_WORKER
+        .may_load(storage, (&service_name, &worker_address))?
+        .ok_or(ContractError::WorkerNotFound)
+}
 
 #[cfg(test)]
 mod tests {
     use super::*;
+    use cosmwasm_std::testing::mock_dependencies;
+    use std::{str::FromStr, vec};
+
+    #[test]
+    fn get_registered_worker_chain_success() {
+        let mut deps = mock_dependencies();
+
+        let worker = Addr::unchecked("worker");
+        let service_name = "validators";
+        let chain_name = ChainName::from_str("ethereum").unwrap();
+        let chains = vec![chain_name.clone()];
+        assert!(register_chain_support(
+            deps.as_mut().storage,
+            service_name.into(),
+            chains,
+            worker.clone()
+        )
+        .is_ok());
+
+        let worker_chains =
+            may_load_chains_per_worker(deps.as_mut().storage, service_name.into(), worker).unwrap();
+        assert!(worker_chains.contains(&chain_name));
+    }
```

Comment:
I think we should have more tests than just this one test. This test only tests registering a single chain, and it doesn't test registering multiple chains in one call, over multiple calls, etc. I'll approve, but can you add a few more tests before merging?

Reply:
Thanks, I added three more tests.